### DISCUSSION TITLE
Recommended way of calling Async Services like IAP Purchase

### DISCUSSION THREAD

**Discussion by TheJoeFin** [microsoft/WindowsAppSDK#2418] [Q&A] [answered]
I am working on adding some In App Purchases to my WinAppSDK app and have be encountering error. I'm sure they are related to `async/await` and windowing problems. I have followed documentation for how to implement `pickers` but the IAP `StoreProduct.RequestPurchaseAsync()` doesn't quite work the same.

Since I will probably encounter similar problems with other UWP APIs, what is the best practice for calling these APIs?

[StoreProduct.RequestPurchaseAsync](https://docs.microsoft.com/en-us/uwp/api/windows.services.store.storeproduct.requestpurchaseasync?view=winrt-22000)

Here is my failed implementation of  using the above API:
https://github.com/TheJoeFin/Simple-Icon-File-Maker/blob/4095db4efdaf633e2a190b45856246ac09f93579/Simple%20Icon%20File%20Maker/Simple%20Icon%20File%20Maker/MainWindow.xaml.cs#L444

**Comment by roxk** [comment, marked as answer]
I haven't used the API, but since you are using WinUI 3 I bet you need to do the same thing desktop bridge app has to do with this API [according to this doc](https://docs.microsoft.com/en-us/windows/uwp/monetize/in-app-purchases-and-trials#desktop). I reached this doc by looking at the exception section in the doc you linked.

If this fixes your issue, the reason why this is needed is that, essentially, the modal normally assumes it's in a UWP app and thus in a core window and can show the modal there. A WinUI 3 app has its own window and thus you need to tell the modal that it should display inside that WinUI 3 window instead.

**Reply by TheJoeFin** [reply]
I will try passing the `StoreContext` into the WinRT Interop and see if that does the trick. 

When using these WinRT modals, is there any general rule of what object should be passed into `WinRT.Interop.InitializeWithWindow.Initialize(context, hwnd); ` to get the modal to show properly?
